opalog/opalog.peg

```diff
@@ -5,19 +5,208 @@ package opalog
 // BUGS: the escaped forward solidus (`\/`) is not currently handled for strings.
 //
 
+// currentLocation converts the parser context to a Location object.
 func currentLocation(c *current) *Location {
-    // TODO: Is it possible to propagate  file names into the parser?
+    // TODO(tsandall): is it possible to access the filename from inside the parser?
     return NewLocation(c.text, "", c.pos.line, c.pos.col)
 }
 
+// parseConstantRule attempts to return a rule from a Body.
+// Equality expressions of the form <var> = <ground term> can be
+// converted into rules of the form <var> = <ground term> :- true.
+// This is a concise way of defining constants inside modules.
+// This function handles the conversion.
+func parseConstantRule(stmt Body) (*Rule, error) {
+    if len(stmt) > 1 {
+        return nil, fmt.Errorf("expression must be contained inside rule: %v", stmt)
+    } else if len(stmt) == 1 {
+        stmt := stmt[0]
+        if !stmt.IsEquality() {
+            return nil, fmt.Errorf("non-equality expression must be contained inside rule: %v", stmt)
+        }
+        if !stmt.Terms[2].IsGround() {
+            return nil, fmt.Errorf("constant rule value must be ground: %v", stmt)
+        }
+        switch name := stmt.Terms[1].Value.(type) {
+        case Var:
+            rule := &Rule{
+                Location: stmt.Location,
+                Name: name,
+                Value: stmt.Terms[2],
+                Body: []*Expr{
+                    &Expr{Terms: []*Term{BooleanTerm(true)}},
+                },
+            }
+            return rule, nil
+        default:
+            return nil, fmt.Errorf("rule name must be a variable: %v", stmt)
+        }
+    } else {
+        panic("unreachable")
+    }
+}
+
+}
+
+Module <- _ head:Stmt tail:( ws Stmt )* _ EOF {
+
+    // Check if multiple statements were given.
+    // If a single statement is provided, we return it as is without
+    // creating a Module object. This supports callers that only
+    // require a single statement (e.g., REPL, tests, etc.)
```

Comment:
I like the idea.  I can't tell whether it'll be good to do that with the parser, or whether we should do it outside the parser.  Meaning that if the caller has a string and doesn't know if it represents multiple statements or a single statement, they'll need to check the type of the return value.

An alternative is to have the parser always return a list of statements and then provide a helper function, maybe called parse1, that runs parse, asserts there's only 1 statement, and then returns that 1 statement.